service/history/workflow/task_generator.go

```diff
@@ -827,23 +831,26 @@ func generateSubStateMachineTask(
 		}
 	}
 	// Only set transition count if a task is non-concurrent.
+	var machineLastUpdateVersionedTransition *persistencespb.VersionedTransition
 	transitionCount := int64(0)
-	machineLastUpdateMutableStateTransitionCount := int64(0)
 	if !task.Concurrent() {
+		machineLastUpdateVersionedTransition = node.InternalRepr().GetLastUpdateVersionedTransition()
 		transitionCount = node.InternalRepr().GetTransitionCount()
-		machineLastUpdateMutableStateTransitionCount = node.InternalRepr().GetLastUpdateMutableStateTransitionCount()
+	}
+
+	transitionHistory := mutableState.GetExecutionInfo().TransitionHistory
+	var currentVersionedTransition *persistencespb.VersionedTransition
```

Comment:
Shouldn't this always be `== machineLastUpdateVersionedTransition`? I wonder if we even need MS versioned transition, it would only be relevant if you allow nodes from detached branches in the tree.

Reply:
They are not equal when refreshing tasks.
And I think we still need MS versioned transition, unless we use a different way (a new field) for representing if the ref is concurrent or not. Right now is done by setting lastUpdateVersioned to nil or just the transition count to 0, but neither can be used for mutable state staleness check.

Reply:
> They are not equal when refreshing tasks.

Oh right!


I have an idea for how we'd want to express non-concurrent refs. I think instead of putting it on the ref we can get rid of the `Concurrent()` method on `Task` and force all tasks to implement `Validate()`, then the non-persistent `Ref` can embed the task's `Validate()` function.

We can have a helper that validates no transitions have happened since the ref was created.

Reply:
Synced offline. ^ can be done later as improvement to the HSM framework . Not blocking this PR.